Image map in linked notes

lledieu opened the pull request:
This PR add a way to manage image map in linked notes.

A new button in toolbar is added in m=MOD_NOTES page to initialize the note for image map. The note is typed with TYPE=gallery.
When viewing or editing a gallery note, templates notes_gallery.txt and notes_upd_gallery.txt are used for rendering.
Optional face detection is provided with openCV.

lledieu committed:
Add image map in linked notes

Added a way to display / update linked notes with templates.
- when TYPE=<value> is found in notes, using template notes_<value>.txt
for display and notes_upd_<value>.txt for update
- Added option ajax=on for m=NOTES and m=MOD_NOTES in order to get data
from templates

Sample note content:
TITLE=Page title
TYPE=gallery
{"title":"Page title","path":"doc","img":"image.jpg",map:[
{"fn":"ludovic","sn":"ledieu","oc":"0","alt":"Ludovic
LEDIEU","shape":"rect","coords","186,229,294,344"}
]}

Path may have :
doc for %prefix;m=DOC&s=
private for %bvar.gallery_path_private;
public for %bvar.gallery_path;
any other value like http://mon.site/

Using jquery.maphilight to ease highlighting
https://github.com/kemayo/maphilight

Using OpenCV for face detection :
https://docs.opencv.org/3.4/d2/d99/tutorial_js_face_detection.html
(only working with doc path)

Added a way to load xml file (needed for face detection librabry)

Missing file + easier to init

Template notes_upd_gallery.txt was missing

Ease gallery note initialisation from toolbar.

lledieu committed:
Rework after feedback

Added a way to list only gallery links :
i=<n>&m=LINKED&type=gallery

Improved templates :
- Better use of firstname and surname
- it's possible a link to misc notes, any geneweb link, external linked
- better JSON robustess. Missing fields are not blocking.

lledieu committed:
Description + fixes

Using variant of safe_html for security.
FIXME: Problem with tag attributes " should be \"

Added description field.
FIXME: It is not possible to %s

Added missing %hidden; in form.

Added placeholder in href field.

Fixed wrong selection.
No new line if selection is less than 5px side.

lledieu committed:
More fixes

lledieu committed:
Remove useless width

lledieu committed:
Go back without Markup + fixes

lledieu committed:
util.mli was wrong !

a2line committed:
mise en forme

lledieu committed:
Security is back

Using Yojson to parse JSON for safe_html and sting_with_macros where it
is needed.

Face detection is now optional. To use it, set in gwf file :
face_detection=on

Edit button was broken in A2 commit.

dataTables.bootstrap4.min.js is commented to remove an erreur on
console.

lledieu committed:
Fixes

a2line commented:
TODO if we merge this: AppVeyor is missing yojson.

lledieu commented:
> TODO if we merge this: AppVeyor is missing yojson.

Je ne me faisais pas de soucis puisque Yojson est déjà mentionné par l'API...

a2line commented:
L'api n'est pas construite sur Appveyor. 

À l'inverse Travis plante sur de le fait que yojson est défini deux fois…

a2line committed:
Même mise en forme sur le modal en toolbar

a2line committed:
Close modal on validation

a2line commented:
J’ai cette trace qui revient souvent sur cette branche :
```
Backtrace:
Raised by primitive operation at file "unix.ml", line 195, characters 7-33
Called from file "lib/wserver/wserver.ml", line 439, characters 20-53
Called from file "lib/wserver/wserver.ml", line 443, characters 4-11
Called from file "lib/wserver/wserver.ml", line 520, characters 6-37
Called from file "lib/wserver/wserver.ml", line 588, characters 18-57
```

sagotch commented on lib/notes.ml:
```diff
@@ -390,24 +440,65 @@ let print conf base =
+      in
+      match templ with
+      | Some ic ->
+         begin match p_getenv conf.env "ajax" with
```
Pour moi, `"ajax"` n'est pas bien choisi (on pourrait tres bien demander du html en ajax). J'utiliserai plutôt quelque chose comme`output=json`.

Pour ce qui est du coding style, le pattern matching n'apporte pas grand chose ici, j'utiliserai juste un `if =`
```ocaml
if p_getenv conf.env "output" = Some "json"
then begin ... end
else begin
  let title = try List.assoc "TITLE" nenv with Not_found -> "" in
  let title = Util.safe_html title in
  match p_getint conf.env "v" with
  | Some cnt0 -> print_notes_part conf base fnotes title s cnt0
  | None -> print_whole_notes conf base fnotes title s None
end
```

Ensuite, `ic` ne sert à rien dans le `"ajax=on"`, donc inutile d'ouvrir le fichier avant.

lledieu replied:
Au début j'avais mis `json=on` mais j'ai remplacé par `ajax=on` pour ne pas mentionner le format et rester générique. Le type gallery utilise le format JSON mais tout format serait possible à condition de gérer le header en accord.

Effectivement, pour `ic`, je peux mieux faire. N'ayant pas trouvé de fonction existante plus adapté, je me suis accommodé avec ;)

sagotch replied:
CF mon message sur l'API. Ce parametre n'aurait plus de sens, et l'api genre deja `output=pb`, `output=xml` et `ouput=json` automatiquement.

lledieu replied:
Il faut que j'arrive à installer protobuf avant de pouvoir jouer avec l'API...

sagotch replied:
Ah oui, ça c'est un autre problème... :(

lledieu replied:
Il est préférable d'utiliser Yojson si protobuf (API) n'est pas suffisamment standard. On peut éventuellement se rapprocher au maximum du fonctionnement de l'API dans l'espoir de pouvoir converger plus facilement le jour venu...

sagotch commented on lib/notes.ml:
```diff
@@ -390,24 +440,65 @@ let print conf base =
+      | Some ic ->
+         begin match p_getenv conf.env "ajax" with
+         | Some "on" ->
+            let charset = if conf.charset = "" then "utf-8" else conf.charset in
```
Utiliser cette fonction (qui est très mal nommée) ?

```
Util.html ~content_type:"application/json" conf ;
```

lledieu replied:
Je vais créer Util.header_json conf qui sera plus explicite...

sagotch commented on lib/notes.ml:
```diff
@@ -390,24 +440,65 @@ let print conf base =
+  let typ = try List.assoc "TYPE" env with Not_found -> "" in
+  let templ =
+    let fname = "notes_upd_" ^ typ in
+    Util.open_templ conf fname
```
Il ne faut pas chercher a ouvrir le fichier ici car il est potentiellement inutile.

lledieu commented:
> J’ai cette trace qui revient souvent sur cette branche

À lire le code associé, on a l'impression que l'exécution d'une requête est interrompue par le client. Il va falloir que tu explicites à quel moment ça se produit... Par exemple, quelle est la requête qu'il est en train de traiter ?

sagotch commented on lib/notes.ml:
```diff
@@ -390,24 +440,65 @@ let print conf base =
   in
-  let (env, s) = read_notes base fnotes in
-  Wiki.print_mod_view_page conf true "NOTES" fnotes title env s
+  match templ, p_getenv conf.env "notmpl" with
```
Cela "casse" le code actuel, qui n'avait pas besoin d'une variable supplémentaire (et templ ne devrait pas etre defini - cf commentaire plus haut).

lledieu replied:
J'ai du mal à comprendre cette remarque...
Lorsque la note est typée gallery, le template notes_upd_gallery.txt prend en charge le m=MOD_NOTES. Au cas où le JSON retourné avec l'option ajax=on (ou format=json) serait incorrect, on arrive dans une situation de blocage. L'option notempl=on permet alors de forcer un m=MOD_NOTES sans l'utilisation de template afin de pouvoir corriger directement le JSON en mode texte (area)... À priori cette option ne devrait servir que très rarement ; je n'en ai eu besoin qu'au début du développement.

sagotch commented on lib/notes.ml:
```diff
@@ -390,24 +440,65 @@ let print conf base =
+         | _ -> Templ.copy_from_templ conf [] ic
+         end
+      | None ->
+         let title = try List.assoc "TITLE" nenv with Not_found -> "" in
```
C'est (vraiment) pas grand chose, mais ceci économise (parfois) un appel de fonction.
```
let title = try Util.safe_html (List.assoc "TITLE" nenv) with Not_found -> "" in
```

a2line committed:
Merge branch 'master' into imgmap

sagotch commented on lib/notes.ml:
```diff
@@ -390,24 +440,65 @@ let print conf base =
     | None -> ""
   in
+  let (env, s) = read_notes base fnotes in
+  let typ = try List.assoc "TYPE" env with Not_found -> "" in
```
Une bonne raison d'utiliser `TYPE` et non `TEMPLATE` directement ? Autant nommer les choses par leur nom.

lledieu replied:
L'utilisateur n'a pas à savoir que c'est géré par un template ou par GeneWeb. Ça pourrait changer à tout moment. Ce qui intéresse l'utilisateur, c'est de savoir que cette note est de type gallery. C'est pour ça que j'avais utilisé TYPE.

sagotch reviewed (changes requested):
J'aimerai bien un/des screenshot(s) du resultat, si possible.

Sinon j'ai fais pas mal de commentaire, plus sur la forme que sur le fond.

sagotch commented on lib/notes.ml:
```diff
@@ -380,6 +397,39 @@ let print_what_links conf base fnotes =
   end;
   Hutil.trailer conf
 
+let safe_gallery conf s =
```
En y repensant, on a déjà une API capable de retourner sur json (en ajoutant le parametre output=json ou lieu de output=pb). L'API vaut ce qu'elle vaut, mais en tout cas elle est là, et bien fournie.

Du coup, ce qui est fait dans safe_gallery est en fait juste un boulot pour l'api.

Il fautdrait ajouter donc un mode (`API_GALLERY` ?), qui utiliserait les mecanismes qui existent déjà, sans quoi on aura deux API à maintenir. J'aurai préféré utilser uniquement Yojson et ne pas passer par ces fichiers proto, mais l'api actuelle est bien installée, et s'il est tout à fait envisageable de la refaire avec Yojson, c'est beaucoup de boulot.

lledieu replied:
Ici c'est un peu particulier, j'ai à la fois du JSON et du HTML. Je voulais profiter des fonctions existantes safe_html et string_with_macros pour les morceaux de HTML présents dans le JSON mais ça cassait le JSON. J'ai donc été obligé de pacourrir le JSON pour appliquer ces fonctions uniquement là où elles étaient nécessaires.

sagotch replied:
Tu me dis si je me trompe, mais il n'y a que du json ici (qui contient peut etre du html, ou pas, comme chaine de charactere dans certains champs, mais c'est une chaine de caracteres json).

En tout cas, ce n'est pas une page HTML qui embarque du json, (c'est d'ailleurs pour ça que le content type est mis à `application/json`), ça retourne des données brutes au format json.

C'est typiquement une utilisation d'une API geneweb.

a2line commented:
On créé une note, dont on initialise la syntaxe json avec le dernier bouton de la barre d'outil des notes :
![image](https://user-images.githubusercontent.com/10353895/59505765-ab3b0d00-8ea6-11e9-90a5-90605597a9dc.png)

On l'édite :
![image](https://user-images.githubusercontent.com/10353895/59505814-d160ad00-8ea6-11e9-97e7-ce4f92e9dcd8.png)

![image](https://user-images.githubusercontent.com/10353895/59505844-edfce500-8ea6-11e9-9827-0a62d85cb1c8.png)

On ajoute des meta-données (quatre types possible) :
![image](https://user-images.githubusercontent.com/10353895/59506098-ad519b80-8ea7-11e9-905e-b76f3ec20bd1.png)

On sauvegarde la note :
![image](https://user-images.githubusercontent.com/10353895/59506116-bcd0e480-8ea7-11e9-9e7a-0dcc42e0662e.png)

Puis on visualise la note :
![image](https://user-images.githubusercontent.com/10353895/59506162-d7a35900-8ea7-11e9-91ac-36faeb1e04be.png)

lledieu commented on lib/notes.ml:
```diff
@@ -390,24 +440,65 @@ let print conf base =
+            Wserver.header "Content-type: application/json; charset=%s" charset ;
+            Wserver.printf "%s" (match typ with
+              | "gallery" -> safe_gallery conf s
+              | _ -> s 
```
Oui, ce n'est pas assez générique. Reste à savoir quel est le bon format par défaut...

sagotch replied:
De base, les notes retournent du html.

lledieu committed:
Several improvements

On person page :
- remove link to linked pages in notes module
- linked pages are shown in menubar
- separate standard linked page / gallery linked page
- add a counter for linked pages

LINKED :
- when type=gallery, display all linked gallery pages with image preview
- when type=, display non typed linked notes
- without type param, all linked notes are listed

gallery display :
- Added RLM link
- fixes :
- remove unwanted d-inline in legend
- do not display twice comment when first_name and surname missing

gallery editor :
- get back scroll on table
- added return false to fix brocken face detection
- hilight was missing in legend (missing <span>)
- try to remove unwanted space in legend (TODO: one is still present)

menubar :
- xl size is needed

N.B.: Use Ctrl+F5 to force refresh CSS on first use

hgouraud commented:
Tu n'as pas intégré la proposition "group" que j'ai faite dans imgmap-hg?

hgouraud commented:
Le texte du champ description devrait passer par un traitement "string_with_macro" (au minimum %s)

lledieu commented:
> Tu n'as pas intégré la proposition "group" que j'ai faite dans imgmap-hg?

Non, j'ai besoin de bien comprendre l'intérêt avant. Le formulaire est déjà bien chargé pour un utilisateur lambda...

lledieu commented:
> Le texte du champ description devrait passer par un traitement "string_with_macro" (au minimum %s)

C'est déjà fait dans safe_gallery, non ?
https://github.com/lledieu/geneweb/blob/d81dfcf64138e48e5487e92a4c02aa54eb8f1b3e/lib/notes.ml#L459

hgouraud commented:
Pourquoi répéter First_name Surname (First_Name Surname) si c'est la même chose?

Pour la notion de groupe, je pense que la surcharge sur le formulaire de saisie est minime (une petite colonne). La fonction a été prévue dans maphighlight. Ca colle avec ce que je souhaitais (pouvoir identifier dans une photo de groupe des paires, des fratries, ...). La seule idée complémentaire serait de rendre cela optionnel ou modal (presser une touche supplémentaire ou cliquer sur un bouton pour l'activer). Sur l'exemple donné dans maphighlight, il me semble qu'il y a un mécanisme de ce genre que j'ai du mal à décoder!

hgouraud commented:
> > Le texte du champ description devrait passer par un traitement "string_with_macro" (au minimum %s)
> 
> C'est déjà fait dans safe_gallery, non ?
> https://github.com/lledieu/geneweb/blob/d81dfcf64138e48e5487e92a4c02aa54eb8f1b3e/lib/notes.ml#L459

Je ne crois pas que ca marche!

lledieu committed:
Fix : safe_gallery was never called

lledieu commented:
> Je ne crois pas que ca marche!

Effectivement ! Ça doit être mieux maintenant...

lledieu commented:
> Pourquoi répéter First_name Surname (First_Name Surname) si c'est la même chose?

Peux-tu préciser dans quel contexte tu as ça ?

hgouraud commented:
> > Pourquoi répéter First_name Surname (First_Name Surname) si c'est la même chose?
> 
> Peux-tu préciser dans quel contexte tu as ça ?

En bas de l'image, en mode "affichage", 
Sur cette image : Tree  Guillaume Weeger (Guillaume Weeger)Henri Gouraud (Henri Gouraud)Alexandre Fontana (Alexandre Fontana) ...

lledieu commented:
> > > Pourquoi répéter First_name Surname (First_Name Surname) si c'est la même chose?
> > 
> > 
> > Peux-tu préciser dans quel contexte tu as ça ?
> 
> En bas de l'image, en mode "affichage",
> Sur cette image : Tree Guillaume Weeger (Guillaume Weeger)Henri Gouraud (Henri Gouraud)Alexandre Fontana (Alexandre Fontana) ...

Je n'observe pas ce comportement. Tu as quoi dans le champ commentaire ?

hgouraud commented:
> > > > Pourquoi répéter First_name Surname (First_Name Surname) si c'est la même chose?
> > > 
> > > 
> > > Peux-tu préciser dans quel contexte tu as ça ?
> > 
> > 
> > En bas de l'image, en mode "affichage",
> > Sur cette image : Tree Guillaume Weeger (Guillaume Weeger)Henri Gouraud (Henri Gouraud)Alexandre Fontana (Alexandre Fontana) ...
> 
> Je n'observe pas ce comportement. Tu as quoi dans le champ commentaire ?

Ok, effectivement cette info était répétée dans le champ commentaire. je l'ai supprimée et ca colle.

a2line commented:
Idée : si l’on place les boutons à la fin de la menubar, on peut les masquer quand ils sont à 0 (n'ont aucun intérêt, d'ailleurs l’actuel « pages liées » textuel n'est pas affiché que quand il n’y a pas de notes liées), et ce sans perturber la navigation avec un décalage des boutons dans la barre. Ceci dit, j’aime bien aussi ces deux boutons à côté de la fiche principale…

lledieu commented:
> Idée : si l’on place les boutons à la fin de la menubar, on peut les masquer quand ils sont à 0

Quelle que soit leur position, on peut les masquer. C'est juste que j'avais pris l'habitude de toujours les avoir car je les avais initiés pour mon interfaçage PHP et que GeneWeb ne pouvait pas savoir que le compteur était à 0...

a2line commented:
Oui mais c'est plus perturbant s'ils sont en début de barre car les boutons vont se décaler d'une fiche à l'autre fonction de la présence de notes/imagemaps ou non.

a2line commented:
Salut @lledieu ! Est-ce que tu aurais travaillé à nouveau sur cet outil de ton côté depuis ces discussions ? 

Pour @canonici et @Swrup l'outil ajoute du code json dans les notes pour permettre d'enrichir les images qui y sont appelées (cf. mes aperçus plus haut). Il ne touche pas aux bases, ni au format des bases, seulement aux notes (.txt auxiliaires). Les modifs .ml sont quasi que des ajouts et le rebasage doit être assez facile. Voir les discussions de Sagotch sur la présence d'un support préexistant pour les req. json dans l'API mais il me semble que l'API a entièrement été dépréciée (ou fusionnée ?) depuis. En gros je vois à relire :
* ajout du support des .xml dans gwdaemon
* ajout d'un support .json pour ces imagemaps dans notes.ml/perso.ml
* le reste c'est tout le js/css/txt pour le template, ie. la partie visible déjà illustrée dans ce fil plus haut.

Côté template, deux petits ponts des notes liées plus ce nouveau concept d’« images liées » inhérents sont intégrés dans la barre de menu. Je ne sais plus si c'est dans la PR, mais c'est en prod depuis sur la base perso de Ludovic : menubar est étendue de deux boutons pour mettre en avant les notes liées à l'individu avec un comptage en exposant, de même à côté pour les images liées (ici pas d'imagemap liée mais vous voyez bien l'idée). Il est vrai que upd_nldb.exe qui va lier les notes aux fiches laisse une trace assez peu en évidence au milieu des notes et je trouve ces ajouts pertinents pour montrer qu'il y a d'autres sources sur les fiches qu'un simple bouton flottant un peu perdu.
![image](https://user-images.githubusercontent.com/10353895/191317964-0857a7f7-a603-4031-a781-6e1d09c59c70.png) 
On discutait de la position en menubar de ces boutons là sur les derniers échanges, mais ce n'est pas important du tout, l'outil était fonctionnel en quelques jours et ne demande qu'à être intégré et amélioré depuis. J'ajoute enfin qu'@hgouraud avait proposé une amélioration pour ses besoins qui n'est pas là.

lledieu commented:
Bonjour A2,
Je n'ai rien changé de mon côté. À noter que le système que j'avais expérimenté pour détecter automatiquement les visages n'est pas stable. Le reste fonctionne toujours. Vous pouvez reprendre ce qui vous intéresse là dedans et l'enrichir à votre guise. Si vous avez besoin d'explications techniques, je peux répondre aux questions.
